File: Backend/app/services/nasa_media.py
"""NASA Image & Video Library API client."""

from typing import List, Dict, Tuple
import time

import httpx

from app.config import NASA_IMAGES_API_URL

MEDIA_CACHE: Dict[str, Tuple[float, dict]] = {}
CACHE_TTL = 3600  # 1 hour



async def search_nasa_media(
    query: str = "space",
    media_type: str | None = None,
    page: int = 1,
    page_size: int = 20,
) -> dict:
    """
    Search the NASA Image & Video Library.
    Returns parsed dict with 'items' list and 'total_hits'.
    """
    cache_key = f"{query}_{media_type}_{page}_{page_size}"
    now = time.time()
    
    # Return from cache if valid
    if cache_key in MEDIA_CACHE:
        timestamp, cached_data = MEDIA_CACHE[cache_key]
        if now - timestamp < CACHE_TTL:
            return cached_data

    params = {
        "q": query,
        "page": page,
        "page_size": page_size,
    }
    if media_type and media_type in ("image", "video"):
        params["media_type"] = media_type

    try:
        async with httpx.AsyncClient(timeout=15.0) as client:
            resp = await client.get(
                f"{NASA_IMAGES_API_URL}/search", params=params
            )
            resp.raise_for_status()
            data = resp.json()

            collection = data.get("collection", {})
            raw_items = collection.get("items", [])
            total_hits = collection.get("metadata", {}).get("total_hits", 0)

            items: List[dict] = []
            for item in raw_items:
                item_data = item.get("data", [{}])[0] if item.get("data") else {}
                links = item.get("links", [])
                thumbnail = ""
                for link in links:
                    if link.get("rel") == "preview":
                        thumbnail = link.get("href", "")
                        break

                nasa_id = item_data.get("nasa_id", "")
                m_type = item_data.get("media_type", "image")

                # For videos, construct preview URL (resolved on-demand when user clicks)
                video_url = None
                if m_type == "video" and nasa_id:
                    video_url = f"{NASA_IMAGES_API_URL}/asset/{nasa_id}"

                items.append({
                    "nasa_id": nasa_id,
                    "title": item_data.get("title", "Untitled"),
                    "description": (item_data.get("description") or "")[:300],
                    "media_type": m_type,
                    "thumbnail": thumbnail,
                    "video_url": video_url,
                    "date_created": item_data.get("date_created", ""),
                    "center": item_data.get("center", ""),
                })

        result = {"items": items, "total_hits": total_hits}
        MEDIA_CACHE[cache_key] = (now, result)
        return result

    except httpx.HTTPError as exc:
        print(f"[NASA Media] Error: {exc}")
        return {"items": [], "total_hits": 0}

from typing import Optional


async def resolve_video_url(nasa_id: str) -> Optional[str]:
    """Resolve the actual mp4 URL for a video by its NASA ID."""
    try:
        async with httpx.AsyncClient(timeout=15.0) as client:
            resp = await client.get(f"{NASA_IMAGES_API_URL}/asset/{nasa_id}")
            if resp.status_code == 200:
                data = resp.json()
                items = data.get("collection", {}).get("items", [])
                for item in items:
                    href = item.get("href", "")
                    if href.endswith(".mp4") and "~orig" in href:
                        return href
                for item in items:
                    href = item.get("href", "")
                    if href.endswith(".mp4"):
                        return href
    except Exception:
        pass
    return None

File: Backend/src/services/nasaMedia.js
const axios = require('axios');

const NASA_IMAGES_API_URL = 'https://images-api.nasa.gov';

// In-memory cache with TTL
const mediaCache = {};
const CACHE_TTL = 3600 * 1000; // 1 hour in ms

/**
 * Search NASA Image & Video Library.
 */
async function searchNasaMedia(query = 'space', mediaType = null, page = 1, pageSize = 20) {
    // Rotate default query for variety
    if (query.trim().toLowerCase() === 'space') {
        const dailyKeywords = ['nebula', 'galaxy', 'supernova', 'james webb', 'aurora', 'black hole', 'exoplanet', 'milky way', 'hubble'];
        const dayOfYear = Math.floor((Date.now() - new Date(new Date().getFullYear(), 0, 0)) / (1000 * 60 * 60 * 24));
        query = dailyKeywords[dayOfYear % dailyKeywords.length];
    }

    const cacheKey = `${query}_${mediaType}_${page}_${pageSize}`;
    const now = Date.now();

    // Return from cache if valid
    if (mediaCache[cacheKey] && (now - mediaCache[cacheKey].timestamp) < CACHE_TTL) {
        return mediaCache[cacheKey].data;
    }

    const params = { q: query, page, page_size: pageSize };
    if (mediaType && ['image', 'video'].includes(mediaType)) {
        params.media_type = mediaType;
    }

    try {
        const resp = await axios.get(`${NASA_IMAGES_API_URL}/search`, { params, timeout: 15000 });
        const collection = resp.data.collection || {};
        const rawItems = collection.items || [];
        const totalHits = (collection.metadata || {}).total_hits || 0;

        const items = rawItems.map((item, idx) => {
            const itemData = (item.data && item.data[0]) || {};
            const links = item.links || [];

            let thumbnail = '';
            for (const link of links) {
                if (link.rel === 'preview') {
                    thumbnail = (link.href || '').replace('http://', 'https://');
                    break;
                }
            }

            const nasaId = itemData.nasa_id || '';
            const mType = itemData.media_type || 'image';
            let videoUrl = null;
            if (mType === 'video' && nasaId) {
                videoUrl = `${NASA_IMAGES_API_URL}/asset/${nasaId}`;
            }

            return {
                nasa_id: nasaId,
                title: itemData.title || 'Untitled',
                description: (itemData.description || '').substring(0, 300),
                media_type: mType,
                thumbnail,
                video_url: videoUrl,
                date_created: itemData.date_created || '',
                center: itemData.center || '',
            };
        });

        const result = { items, total_hits: totalHits };
        mediaCache[cacheKey] = { timestamp: now, data: result };
        return result;

    } catch (err) {
        console.error(`[NASA Media] Error: ${err.message}`);
        return { items: [], total_hits: 0 };
    }
}

/**
 * Resolve the actual mp4 URL for a NASA video by its ID.
 */
async function resolveVideoUrl(nasaId) {
    try {
        const resp = await axios.get(`${NASA_IMAGES_API_URL}/asset/${nasaId}`, { timeout: 15000 });
        const items = (resp.data.collection || {}).items || [];

        const preferences = ['~large.mp4', '~medium.mp4', '~mobile.mp4', '~orig.mp4'];
        const mp4s = items
            .map(i => i.href || '')
            .filter(href => href.endsWith('.mp4'));

        let bestUrl = null;
        for (const pref of preferences) {
            for (const url of mp4s) {
                if (url.endsWith(pref)) {
                    bestUrl = url;
                    break;
                }
            }
            if (bestUrl) break;
        }

        if (!bestUrl && mp4s.length > 0) {
            bestUrl = mp4s[0];
        }

        if (bestUrl) {
            return bestUrl.replace('http://', 'https://');
        }
    } catch {
        // silently fail
    }
    return null;
}

module.exports = { searchNasaMedia, resolveVideoUrl };
